---
title: Applied Statistical Methods - `r tools::toTitleCase(params$doctype)` 11
author: Peter von Rohr
date: 2022-05-22
output:
  bookdown::pdf_document2:
    toc: false
    number_sections: false
    keep_md: false
    keep_tex: false
header-includes:
 \usepackage{longtable}
 \usepackage{float}
params:
  doctype:
    label: 'Document Type'
    value: 'solution'
    choices: ['exercise', 'solution', 'notebook']
  isonline:
    label: 'Online (y/n)'
    value: TRUE
    choices: [TRUE, FALSE]
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Problem 1: Marker Effects Model
```{r ex11p01-setup, echo=FALSE}
s_ex11_p01_data_path <- "https://charlotte-ngs.github.io/asmss2022/data/asm_geno_sim_data.csv"
if (!params$isonline)
  s_ex11_p01_data_path <- file.path(here::here(), "docs", "data", "asm_geno_sim_data.csv")

sigma_q2 <- 3
sigma_e2 <- 36
```

Predict genomic breeding values using a marker effects model. The dataset is available from

```{r, echo=FALSE}
cat(s_ex11_p01_data_path, "\n")
```

### Hints

* The variance $\sigma_q^2$ of the marker effect is $`r sigma_q2`$.
* The residual variance $\sigma_e^2$ is $`r sigma_e2`$
* The sex of each animal can be modelled as a fixed effect

<!-- your-solution-start
### Your Solution

---  your-solution-end -->


<!-- master-solution-start -->

### Solution

<!-- master-solution-end -->


## Problem 2: Breeding Value Based Model
```{r ex11p02-setup, echo=FALSE}
s_ex11_p02_data_path <- "https://charlotte-ngs.github.io/asmss2022/data/asm_geno_sim_data.csv"
if (!params$isonline)
  s_ex11_p02_data_path <- file.path(here::here(), "docs", "data", "asm_geno_sim_data.csv")

sigma_q2 <- 3
sigma_g2 <- 3*sigma_q2
sigma_e2 <- 36
```

Use the same dataset as in Problem 1 to predict genomic breeding values based on a breeding-value model. The dataset is available from

```{r, echo=FALSE}
cat(s_ex11_p02_data_path, "\n")
```

### Hints

* The genomic variance $\sigma_g^2$ of the marker effect is $`r sigma_g2`$.
* The residual variance $\sigma_e^2$ is $`r sigma_e2`$
* The sex of each animal can be modelled as a fixed effect

Use the following function to compute the genomic relationship matrix $G$ based on the matrix of genotypes

```{r}
computeMatGrm <- function(pmatData) {
  matData <- pmatData
  # check the coding, if matData is -1, 0, 1 coded, then add 1 to get to 0, 1, 2 coding
  if (min(matData) < 0) matData <- matData + 1
  # Allele frequencies, column vector of P and sum of frequency products
  freq <- apply(matData, 2, mean) / 2
  P <- 2 * (freq - 0.5)
  sumpq <- sum(freq*(1-freq))
  # Changing the coding from (0,1,2) to (-1,0,1) and subtract matrix P
  Z <- matData - 1 - matrix(P, nrow = nrow(matData), 
                             ncol = ncol(matData), 
                             byrow = TRUE)
  # Z%*%Zt is replaced by tcrossprod(Z)
  return(tcrossprod(Z)/(2*sumpq))
}
```

<!-- your-solution-start
### Your Solution

---  your-solution-end -->


<!-- master-solution-start -->

### Solution

<!-- master-solution-end -->


## Problem 3

<!-- your-solution-start
### Your Solution

---  your-solution-end -->


<!-- master-solution-start -->

### Solution

<!-- master-solution-end -->
 
