---
title: Applied Statistical Methods - `r tools::toTitleCase(params$doctype)` 3
author: Peter von Rohr
date: 2022-03-12
output:
  bookdown::pdf_document2:
    toc: false
    number_sections: false
    keep_md: false
    keep_tex: false
header-includes:
 \usepackage{longtable}
 \usepackage{float}
params:
  doctype:
    label: 'Document Type'
    value: 'solution'
    choices: ['exercise', 'solution', 'notebook']
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r ex03p01-setup, echo=FALSE}
s_ex03p01_data_path <- "https://charlotte-ngs.github.io/asmss2022/data/asm_flem_genomic_data.csv"
```

## Problem 1: Linear Regression on Genomic Information
Use the following dataset which is also given in: 

`r  s_ex03p01_data_path`

to estimate marker effects for the single loci using a linear regression model.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
tbl_ex03p01_data <- readr::read_csv(file = s_ex03p01_data_path)
knitr::kable(tbl_ex03p01_data,
             booktabs = TRUE,
             longtable = FALSE,
             escape = FALSE)
```


<!-- your-solution-start
### Your Solution

* Read the data using `read.csv()`

* Re-code the genotypes to numeric values

* Fit the multiple regression to the data


---  your-solution-end -->


<!-- master-solution-start -->

### Solution

* Read the data using `read.csv()`

```{r}
s_ex03p01_data_path <- "https://charlotte-ngs.github.io/asmss2022/data/asm_flem_genomic_data.csv"
tbl_ex03p01_data <- readr::read_csv(file = s_ex03p01_data_path)
tbl_ex03p01_data
```

* Re-code the genotypes to numeric values

```{r}
tbl_ex03p01_data <- dplyr::mutate(tbl_ex03p01_data, 
                                  `SNP G` = dplyr::recode(`SNP G`,
                                                          "$G_2G_2$" = 0, 
                                                          "$G_1G_2$" = 1,
                                                          "$G_1G_1$" = 2))
tbl_ex03p01_data <- dplyr::mutate(tbl_ex03p01_data, 
                                  `SNP H` = dplyr::recode(`SNP H`,
                                                          "$H_2H_2$" = 0, 
                                                          "$H_1H_2$" = 1,
                                                          "$H_1H_1$" = 2))

tbl_ex03p01_data
```

* Fit the multiple regression to the data

```{r}
lm_mult_reg_genomic <- lm(formula = Observation ~ `SNP G` + `SNP H`, data = tbl_ex03p01_data)
summary(lm_mult_reg_genomic)
```

The marker-effects for the two loci correspond to

```{r}
vec_coeff <- coefficients(lm_mult_reg_genomic)
cat("\nMarker Effect for locus G: ", vec_coeff["`SNP G`"], "\n")
cat("Marker Effect for locus H: ", vec_coeff["`SNP H`"], "\n")
```

<!-- master-solution-end -->


## Problem 2: Regression On Dummy Variables
Use the dataset with the breeds assigned to every animal and find out the influence of the breed on the response variable `body weight`. The data is available from

```{r echo=FALSE}
s_ex03p02_data_path <- "https://charlotte-ngs.github.io/asmss2022/data/asm_bw_flem.csv"
s_ex03p02_data_path
```

Start by fitting a linear model with `Breed` as the only factor in the model, hence ignore the independent variables such as `Breast Circumference`, `BCS` and `HEI`. 

<!-- your-solution-start
### Your Solution

* Read the data

* Fit a linear model with breed as the only factor


---  your-solution-end -->


<!-- master-solution-start -->

### Solution

* Read the data

```{r}
s_ex03p02_data_path <- "https://charlotte-ngs.github.io/asmss2022/data/asm_bw_flem.csv"
tbl_ex03p02_data <- readr::read_csv(file = s_ex03p02_data_path)
tbl_ex03p02_data
```

* Fit a linear model including breed as a factor

```{r}
lm_reg_dummy_bw_breed <- lm(formula = `Body Weight` ~ Breed, data = tbl_ex03p02_data)
summary(lm_reg_dummy_bw_breed)
```

<!-- master-solution-end -->


## Problem 3: Estimable Function
Use the matrix vector-notation to setup the model for a regression on dummy variable with the data on breeds and body weight as used in Problem 2. The aim of this problem is to find the estimable functions used in the output of `lm()`. 

The model is given by 

$$\mathbf{y} = \mathbf{Xb} + \mathbf{e}$$

Setup the least squares normal equations. Find a solution for $\mathbf{b}^0$ and construct the estimable function that is used in the output `lm()`. 

<!-- your-solution-start
### Your Solution

* Define elements of least squares normal equations

* Find a solution for $\mathbf{b}^0$

* Construct the estimable function. As a hint, assume the missing factor level in the output of `lm()` to be zero.

---  your-solution-end -->


<!-- master-solution-start -->

### Solution

<!-- master-solution-end -->
 
